docs: add more option info for application builder (#58569)
angular/angular: adev/src/content/tools/cli/build-system-migration.md

The prebundling process analyzes all the third-party project dependencies within a project and processes them the first time the development server is executed.
This process removes the need to rebuild and bundle the project dependencies both each time a rebuild occurs and each time the development server is executed.

In most cases, no additional customization is required.
However, the prebundling process can be fully disabled or individual dependencies can be excluded if needed by a project.
The `dev-server` builder's `prebundle` option can be used for these customizations.
By default, `prebundle` is set to `true` but can be set to `false` to fully disable prebundling.

<docs-code language="json">
    "serve": {
      "builder": "@angular/build:dev-server",
      "options": {
        "prebundle": false
      },
</docs-code>

To exclude specific dependencies, the `prebundle.exclude` option is also available:

<docs-code language="json">
    "serve": {
      "builder": "@angular/build:dev-server",
      "options": {
        "prebundle": {
          "exclude": ["some-dep"]
        }
      },
</docs-code>

> Some situations where you may need to exclude a dependency include:

Reviewer: Consider: Move this to the start of the section to first tell readers why they might care about this option before explaining how to use it.

> - Customizing loader behavior for imports within the dependency (`loader` option or import attribute)
> - Symlinking a dependency to local code for development

Reviewer: ```suggestion
- Customizing loader behavior for imports within the dependency such as the [`loader` option](#link)
- Symlinking a dependency to local code for development such as [`npm link`](https://docs.npmjs.com/cli/v10/commands/npm-link)
```

Wording: Just removing some unnecessary parentheses and linking to additional context.

> - An error is encountered during prebundling of the dependency

Reviewer: Question: Are there known cases where prebundling needs to be disabled? This feels like a bug we'd want to fix where disabling `prebundling` is only a temporary workaround and not something we'd want to call out in adev if we don't need to.

If you want to keep this, I'd at least rephrase to maintain the verb tense with the other bulleted items:

```suggestion
- Working around an error encountered during prebundling of a dependency
```

Author: Unfortunately, not really. Wording this as a workaround is a good idea though since that's really what it would be.

This option is currently in developer preview.
If you notice any issues, please report them [here](https://github.com/angular/angular-cli/issues).

### Build-time value replacement (define)

The `define` option allows identifiers present in the code to be replaced with another value at build time.
This is similar to the behavior of Webpack's `DefinePlugin` which was previously used with some custom Webpack configurations that used third-party builders.

> The option can either be used within the `angular.json` configuration file or on the commandline.

Reviewer: ```suggestion
The option can either be used within the `angular.json` configuration file or on the command line.
```

Wording: Typically I see "command line" as two words, seems like this comes up a few times in this doc.

> Using the option within the configuration is useful for cases where the values are constant for each build configuration.

Reviewer: ```suggestion
Configuring `define` within `angular.json` is useful for cases where the values are constant and able to be checked in to source control.
```

Wording: Call out source control as a motivator here.

> The commandline usage is preferred for values that may change per build execution such as the git commit hash or an environment variable.

Reviewer: Wording: Unnecessary "the".

```suggestion
Command line usage is preferred for values that may change per build execution such as the git commit hash or an environment variable.
```

> Commandline usage will merge with the configuration usage with the commandline taking precedence if the same identifier is present for both.

Reviewer: Readability: Move this discussion of command line usage after we have explained the `angular.json` usage.

This way we can start with an explanation of `define` in `angular.json` and then add on top of that that it can also be used on the command line as `--define`.

> Commandline usage will merge with the configuration usage with the commandline taking precedence if the same identifier is present for both.

Reviewer: Wording:

```suggestion
The CLI will merge `--define` values from the command line with `define` values from `angular.json`, including both in a build. Command line usage takes precedence if the same identifier is present for both.
```

Reviewer 2: Do we need to call out precedence since it matches how all other CLI-provided options work?

Within the configuration file, the option is in the form of an object.
The keys of the object represent the identifier to replace and the values of the object represent the corresponding replacement value for the identifier.
An example is as follows:

<docs-code language="json">
  "build": {
    "builder": "@angular/build:application",
    "options": {
      ...

>       "define": {
>           "SOME_CONSTANT": "5",
>           "ANOTHER": "'this is a string literal'"
>       }

Reviewer: Maybe this would help lead users towards understanding the difference between the types and the need for extra quotes?

```suggestion
      "define": {
          "SOME_NUMBER": "5",
          "ANOTHER": "'this is a string literal, note the extra single quotes'",
          "REFERENCE": "globalThis.someValue.noteTheAbsentSingleQuotes"
      }
```

    }
  }
</docs-code>

All replacement values are defined as strings within the configuration file.
If the replacement is intended to be an actual string literal, it should be enclosed in single quote marks.

> This allows the flexibility of using any valid JSON type as well as a different identifier as a replacement.

Reviewer: Nit: This feels significant enough to be in an `IMPORTANT` or `HELPFUL` tip. It's way too easy to overlook the extra quotes or think they are a typo.

For commandline usage, the `--define` option uses the format of `IDENTIFIER=VALUE`.
As an example, the following will allow uses the current git commit hash string value in the built application:

<docs-code language="shell">
ng build --define "GIT_HASH='$(git rev-parse --short HEAD)'"
</docs-code>

For non-Windows shells, the quotes around the hash literal can be escaped directly if preferred:

<docs-code language="shell">
ng build --define GIT_HASH=\'$(git rev-parse --short HEAD)\'

> </docs-code>

Reviewer: Consider: I'd just drop the `git` example entirely, I think it's too specific for the level of documentation we're trying to give here.

This option is currently in developer preview.
If you notice any issues, please report them [here](https://github.com/angular/angular-cli/issues).

### Build-time value replacement (define)

The `define` option allows identifiers present in the code to be replaced with another value at build time.
This is similar to the behavior of Webpack's `DefinePlugin` which was previously used with some custom Webpack configurations that used third-party builders.
The option can either be used within the `angular.json` configuration file or on the command line.
Configuring `define` within `angular.json` is useful for cases where the values are constant and able to be checked in to source control.
The command line usage is preferred for values that may change per build execution such as the git commit hash or an environment variable.

> command line usage will merge with the configuration usage with the command line taking precedence if the same identifier is present for both.

Reviewer 2: ```suggestion
Command line usage will merge with the configuration usage with the command line taking precedence if the same identifier is present for both.
```
